### ISSUE TITLE
Tracking: VSTest filtering, FullyQualifiedName translation and selected-test execution

### ISSUE THREAD

**Issue by peterwald** [nunit/nunit3-vs-adapter#505]
This is a tracking (Epic) issue for the recurring family of problems around translating VSTest / MTP filters into NUnit filters, constructing and parsing `FullyQualifiedName`, and keeping test identities stable between discovery and execution. It supersedes the original 2018 scope ("Standardized FullyQualifiedName Support", see https://github.com/Microsoft/vstest-docs/pull/133 and #668).

Symptoms across these issues are usually one of:

- a test is **discovered but never executed** (`No test matches the given testcase filter ...`),
- `Incorrect format for TestCaseFilter` / `Unexpected FQN ... in selection expression` exceptions,
- phantom, never-run entries in Test Explorer,
- the selection is lost on rerun (Azure DevOps / `dotnet test --filter`).

The grouping below is editorial; the native parent/sub-issue links carry progress. Search label: `is:FQNIssue`.

## A. FQN parsing / escaping / special characters

Test is discovered, the filter round-trip mangles the name, the test won't run.

- #1488 - escaped operators `\= \& \| \! \~ \\` discovered but never executed
- #1490 - space before the argument list -> `Unexpected FQN ... in selection expression`
- #1437 - filtering on test name broken in adapter 6.0.0+ (regression)
- #1405 - pipe character `|` in a string argument
- #1377 - tests fail to run after migration to MTP + newest NUnit
- #1349 - MTP filter string includes an unrecognized escape sequence
- #1097 - `).` inside a string argument not detected
- #935 - F# discriminated-union `TestCaseSource` -> `Incorrect format for TestCaseFilter` (External)
- #876 - spaces and special characters unsupported in TestFilter
- #807 - VSTest retry fails with spaces / `&` in a TestCase parameter
- #782 - `IEnumerable` tuple arguments from `TestCaseSource` not run
- #761 - `￿` argument
- #742 - several characters in the name do not work
- #654 - `).` in a TestCase argument breaks Test Explorer grouping

## B. Discovery / execution identity mismatch

Discovery and execution disagree on the test's identity, so the selection cannot be resolved.

- #489 - VS15.6 Real-Time Test Discovery yields unrunnable NUnit tests
- #607 - `No test matches the given testcase filter FQN=...` for `TestCaseData.SetName` names
- #623 - Azure DevOps rerun with `TestFixtureSource` -> `Incorrect format for TestCaseFilter`
- #677 - `TestCaseSource` + `dotnet test --filter` fails (works in VS and without a filter)
- #1408 - array-of-tuple `TestCaseSource` -> phantom entry that never runs
- #1414 - `nunit_random_seed.tmp` not refreshed -> discovery IDs drift -> selection unrunnable

## C. Filter semantics and selection rules

The filter parses fine; the question is what it should select.

- #473 - execute `[Explicit]` tests via the `TestCaseFilter` argument on the command line (design)
- #1117 - `Execution.CheckFilter` behaviour when more than 2000 test cases are selected
- #1197 - `[Explicit]` / `[Ignore]` on a class with a custom `[Category]` ignores the whole category
- #1202 - MSTest-style `ClassName` filter generated on Azure rerun is not understood
- #1257 - `dotnet test --filter` ignored for tests with a `[Platform]` attribute

## Suspected duplicate clusters (needs dedupe)

- One escaping defect: #1488, #1349, #1377, #1405, #807
- The `).` defect: #1097, #654
- Tuple-source enumeration: #782, #1408
- Identity drift: #489, #607, #1414

**Comment by johnmwright** [comment]
Please see https://github.com/nunit/nunit3-vs-adapter/pull/668

**Comment by OsirisTerje** [comment]
### Provisional triage of the sub-issues

Populated this as the filtering / FQN tracker and linked 25 open issues as sub-issues. First-pass classification below - **not verified against the current release yet**; the "needs reconfirmation" ones should be reproduced on the latest adapter (6.x / MTP) before any are actioned or closed.

| Issue | Bucket | Provisional status |
|------|--------|--------------------|
| #1488 | A | Confirmed, likely duplicate of #1405 - likely the canonical escaping defect |
| #1490 | A | Confirmed current |
| #1437 | A | Confirmed regression in 6.0.0+; needs bisect |
| #1349 | A | Confirmed current (MTP) |
| #1405 | A | Reconfirm;  |
| #1377 | A | Reconfirm; likely duplicate of #1349 / #1437 |
| #1097 | A | Reconfirm; pairs with #654 |
| #654  | A | Reconfirm; duplicate candidate of #1097 |
| #935  | A | External (F# DU + VSTest filter format); keep tracked only |
| #876  | A | Reconfirm (2021) |
| #807  | A | Reconfirm; likely duplicate of the escaping cluster |
| #782  | A | Reconfirm; pairs with #1408 |
| #761  | A | Reconfirm (edge char `￿`) |
| #742  | A | Reconfirm (2020); possibly an umbrella duplicate |
| #489  | B | Reconfirm - VS15.6 Real-Time Discovery era, may be obsolete |
| #607  | B | Reconfirm; identity-drift family |
| #623  | B | Reconfirm; Azure rerun, overlaps #935 |
| #677  | B | Reconfirm (2019) |
| #1408 | B | Confirmed current; pairs with #782 |
| #1414 | B | Confirmed current; pairs with #1413 |
| #473  | C | Design question - open by intent |
| #1117 | C | Design / perf question (>2000-case filter cap) |
| #1197 | C | Confirmed current - `[Explicit]`/`[Ignore]` + `[Category]` semantics |
| #1202 | C | Reconfirm; external MSTest `ClassName` filter shape |
| #1257 | C | Investigate; `[Platform]` attribute vs `--filter` |

Suspected duplicate clusters to collapse after reconfirmation:

- Escaping defect: #1488, #1349, #1377, #1405, #807
- `).` in argument: #1097, #654
- Tuple `TestCaseSource` enumeration: #782, #1408
- Discovery/execution identity drift: #489, #607, #1414

Search label `is:FQNIssue` has been applied to all bucket A and B issues; bucket C is tracked through this parent only, since those are filter *semantics* rather than FQN encoding.

**Comment by OsirisTerje** [comment]
### Repro inventory for the sub-issues

Where a reproduction exists and what shape it is. "Repro folder" = the [`nunit3-vs-adapter.issues`](https://github.com/nunit/nunit3-vs-adapter.issues) repo. **None of these have been re-run against the current 6.x / MTP adapter as part of this pass** - the in-repo folders pin old package versions (adapter 4.4.x, NUnit 3.13.x) and the IssueRunner tooling that used to drive them was removed from that repo (commit `5821374`). Verifying "still fails on current" is the next task and needs the folders bumped to current package versions first.

| Issue | Repro available | Location | Notes on current status |
|------|-----------------|----------|-------------------------|
| #1488 | Inline C# + exact `dotnet test` / `dotnet vstest` commands | issue body | Reporter states it reproduces on `main` @ `fe7287f` (2026-09). Self-contained, not independently re-run. |
| #1490 | Inline C# + commands | issue body | Same - reporter says reproduces on `main` @ `fe7287f`. |
| #1437 | External repo | https://github.com/Braedon-Wooding-Displayr/nunit-filter-bug | Regression reported in 6.0.0+. Not re-run. |
| #1405 | Repro folder + inline | `Issue1405/` | Pinned to old adapter. Not re-run. |
| #1377 | Repro folder + `NunitTests.zip` | `Issue1377/` | MTP. Pinned old. Not re-run. |
| #1349 | Repro folder | `Issue1349/` | MTP. Pinned old. Not re-run. |
| #1097 | Repro folder | `Issue1097/` | Pinned old (adapter 4.4.2). Not re-run. |
| #935  | Repro folder + `NunitFSharpTestCaseSource.zip` | `Issue935/` | Labelled `External` (VSTest filter parser). |
| #876  | Shared repro folder | `Issue711/` | The folder's own issue #711 is `closed:fixedinnewversion`; needs a dedicated repro that still fails. |
| #807  | External repo | https://github.com/SabotageAndi/NUnitRetry2213 | SpecFlow retry scenario. Not re-run. |
| #782  | Repro folder + `Sample.zip` | `Issue782/` | Pinned old. Pairs with #1408. Not re-run. |
| #761  | Repro file | `Issue640/Issue761.cs` | Maintainer already doubted this is an FQN issue; likely nunit/nunit#4826 (framework). |
| #742  | Gist + shared repro folder | https://gist.github.com/appel1/579d8a44a9768e9e6e5cf93c57e5d502 , `Issue711/` | Not re-run. Possible umbrella duplicate. |
| #654  | none uploaded | - | johnmwright: same root cause as #622, resolved by PR #668. **Reconfirm / close candidate.** |
| #489  | External repo | https://github.com/Barsonax/Pathfindax | johnmwright: resolved by PR #668. **Reconfirm / close candidate.** |
| #607  | External repo | https://github.com/NetMatch/nunit-adapter-testcase | johnmwright: resolved by PR #668. **Reconfirm / close candidate.** |
| #623  | none uploaded | - | johnmwright: same as #622, resolved by PR #668. **Reconfirm / close candidate.** |
| #677  | External repo | https://github.com/ovebastiansen/nunitadapterissue | Maintainer: FQN with a parameterised class is invalid per the managed-name RFC - likely by-design, not a bug. |
| #1408 | Inline code | issue body; framework PR nunit/nunit#5113 | Fix proposed in framework. Pairs with #782. |
| #1414 | none (detailed narrative only) | - | `nunit_random_seed.tmp` not refreshed. Pairs with #1413. |
| #473  | none (design discussion) | - | Design question, open by intent. |
| #1117 | none in repo; proposed change on a fork | https://github.com/edlund-as/nunit3-vs-adapter/commit/4cf4b5ab0a | `AssemblySelectLimit` (2000) behaviour. Perf/design. |
| #1197 | Framework repro | https://github.com/nunit/nunit.issues/tree/main/Issue4736 | Suspected framework bug, not adapter. |
| #1202 | none | points to nunit/nunit#4651 | External MSTest-style `ClassName` filter. |
| #1257 | External repo | https://github.com/melanchall/drywetmidi (branch `pack-update`) | `[Platform]` attribute vs `--filter`. Not re-run. |

**Immediate follow-ups this surfaces:**

1. #489, #607, #623, #654 were all declared fixed by PR #668 (2020) and never verified or closed - reconfirm against current and close if fixed.
2. #677 looks like invalid-FQN-by-design rather than a bug - needs a maintainer decision.
3. #761, #1197, #1408 have their fix (or root cause) on the framework side - track upstream, not here.
4. The in-repo repro folders (`Issue1097/1349/1377/1405/782/935`) should be bumped to current package versions and re-run to confirm which still fail.

**Comment by OsirisTerje** [comment]
### Spot-check: two repros re-run against the current adapter

Quick CLI re-runs (Windows, .NET 10 SDK, headless - no Visual Studio / real-time discovery). Not a substitute for the full matrix, but two data points:

**#1097** - `Issue1097/` repro bumped to NUnit 4.6.1 + NUnit3TestAdapter 6.3.0, `net8.0`:
`dotnet test` discovers and runs the `TestCase("... (the best, even).")` case (2/2 pass), and `--filter FullyQualifiedName~Test1097` matches it. **Does not reproduce via the CLI** on the current adapter. The VS real-time-discovery path (the original report) is still unverified.

**#1405** - `Issue1405/` repro (MTP project, `net10.0`) with the adapter bumped to 6.3.0:
Running the test app directly (`dotnet Issue1405.dll`) discovers and passes the `TestCase("This | That", false)` case - **the pipe character does not break the MTP runner path on 6.3.0**. `dotnet test` in the headless shell returned a non-specific `error: 1` / zero tests even with the pipe removed, so that path is inconclusive here and needs confirming on a proper VS / `dotnet test` host.

Takeaway: at least #1097 and #1405 may already be fixed or partly fixed on 6.3.0; both need reconfirmation on the VS / `dotnet test` paths before closing. The rest of the sub-issues still need a proper re-run once the repro folders are bumped to current package versions.

**Comment by OsirisTerje** [comment]
### Design note: the two opt-in FQN settings from PR #668 are workarounds, not the fix

`UseParentFQNForParametrizedTests` and `UseNUnitIdforTestCaseId` (added in commit `7635469`, still `false` by default as of 6.3.0) were point workarounds for specific manifestations of the problems tracked here - not a real solution. The adapter's own code comment on `UseParentFQNForParametrizedTests` spells out the cost: it replaces each parameterized case's FQN with the parent method's, so **individual parameterized cases can no longer be selected - running one runs them all**.

Later reports on the linked issues indicate these toggles don't actually resolve the underlying cases for most users. Position: they should be **deprecated and eventually removed** once the real FQN/filter translation is fixed.

Not a hard removal - some users have tuned these to fix their specific setup and rely on them. Suggested path:
1. Mark both as deprecated in the docs and in a discovery-time warning when set to `true`.
2. Keep them working through at least one more major.
3. Remove once the proper fix lands and the tracked issues here are resolved/reconfirmed.

Any real fix should make the default path correct so neither toggle is needed.

**Comment by OsirisTerje** [comment]
The FQN parsing half of this is now written up, with the root cause and a split between what can be fixed without breaking anything and what needs a major version:

- [docs/TestFilterParsing-6.x.md](https://github.com/nunit/nunit3-vs-adapter/blob/main/docs/TestFilterParsing-6.x.md)
- [docs/TestFilterParsing-v7.md](https://github.com/nunit/nunit3-vs-adapter/blob/main/docs/TestFilterParsing-v7.md)

Root cause, in one sentence: our tokenizer splits the filter on `= ~ ! ( ) & |` without consulting a preceding backslash, and then tries to repair the result with a parenthesis-balancing heuristic that reassembles parameterised test names. The test platform's own grammar splits on *unescaped* operators and unescapes the values afterwards, with no notion of an FQN and no balancing. That single difference in layering accounts for **category A in full** — all 13 of them, including #1405, #1488, #1490 and #1501.

It does **not** account for category B (discovery and execution disagreeing on test identity) or category C (filter semantics and selection rules). Those are separate problems and still need their own analysis — worth saying plainly so they do not get quietly assumed fixed when the parser lands.

Credit to @PanzerFowst, whose analysis in #1501 is what pinned this down.

---

*Claude Opus 5 has been assisting in investigating and building these documents.*
